### ISSUE TITLE
Prometheus metric concourse_workers_containers falsely stuck on maximum value

### ISSUE THREAD

**Issue by ari-becker** [concourse/concourse#6956]
## Summary

The Prometheus metric `concourse_workers_containers` was falsely stuck on the maximum value, and not reporting for all workers. Restarting the web Kubernetes pods "unstuck" the reported metric.

## Steps to reproduce

Sorry, no idea! Logs don't seem to show anything.

## Expected results

The correct values reported in real time.

## Actual results

Grafana dashboard showing the stuck metric:
![image](https://user-images.githubusercontent.com/1388505/116545744-73ea7380-a8f9-11eb-80b3-5991435ac751.png)

Querying the exporter directly and seeing the stuck metric:
![image](https://user-images.githubusercontent.com/1388505/116545896-a72d0280-a8f9-11eb-9099-61112dc00500.png)

## Additional context
One of the `web` pods was fresh, the rest were old:
![image](https://user-images.githubusercontent.com/1388505/116546261-173b8880-a8fa-11eb-80fb-a7688e622232.png)

## Triaging info

* Concourse version: 5.8.1
* Browser (if applicable): N/A
* Did this used to work? N/A

**Comment by pn-santos** [comment]
Don't want to highjack but could this be the same as: https://github.com/concourse/concourse/discussions/5992

We still see this behaviour where if a worker switches from ATC1 to ATC2 due to rebalancing then containers of that worker reported by ATC1 will become "stuck" at the last reported value while ATC2 will report the actual container count the worker has.

Obviously this makes it very tricky to properly track (via metrics) the real container counts (right now we are using a query that disregards a count if it's the same over a period of time, which kinda works but is very hacky).

It feels like the ATC never resets a worker's container count after the worker becomes registered with another ATC instance.

**Comment by pn-santos** [comment]
Just to add another data point. We're seeing this happen with the `concourse_workers_tasks` and `concourse_workers_volumes` metrics too. So it seems it's an issue with all the per-worker metrics.

I'd assume that as soon as an (ATC) instance is aware that a worker is no longer registered with it it should reset/zero all it's per-worker metrics to ensure that only the ATC the worker is registered with reports the correct number.

**Comment by taylorsilva** [comment]
Hey, we tried to reproduce this with a docker-compose'd Concourse with two web nodes and one worker. Our theory (based on pn-santos comment) was that when workers rebalanced, thus connecting to the other web node, that the previous web node would not clear out its stale worker metrics. We found out there is code for the prometheus emitter to clear out stale worker stats every 5mins and didn't see the behaviour described in OP: 
https://github.com/concourse/concourse/blob/0874fc5969abc560af526b4c812147871654e154/atc/metric/emitter/prometheus.go#L937-L950

We also didn't see metrics like container count get stuck at a max value, but we didn't explicitly try this.

For now we're out of theories as to what could be causing this. If anyone can share more details about your setup maybe that'll give us a new direction/clue to look at.

**Comment by pn-santos** [comment]
@taylorsilva  thanks for looking into. I can try to provide with info but not sure what might help or be related with this... If you need any config value we are using let me know. 

Here's a 24h sample of our setup for one worker. We have 3 ATCs. First 3 show the metric as reported form each ATC. Last one is all three overlapped.

![ccm0](https://user-images.githubusercontent.com/23340138/123692193-6c3e4100-d84e-11eb-9b1f-ecf69e14fbbc.png)
![ccm1](https://user-images.githubusercontent.com/23340138/123692194-6cd6d780-d84e-11eb-9fd1-c5553b783345.png)
![ccm2](https://user-images.githubusercontent.com/23340138/123692196-6d6f6e00-d84e-11eb-918c-1099c9301a7b.png)
![all](https://user-images.githubusercontent.com/23340138/123692189-6ba5aa80-d84e-11eb-98ed-d75727fe9d92.png)

Here's what I can tell from the graphs:
- Metrics stay "flat" for several hours
- They _sometimes_ reset to zero but most of the time stay flat whenever the worker re-registers with another ATC
- It's pretty obvious from graph that the real value get reported correctly by the ATC that the worker is registered with (I've confirmed the current value using fly and the "real" value is always the ATC the worker is currently registered with)

**Comment by taylorsilva** [comment]
Thanks for the graphs. It very clearly shows the issue with web nodes still reporting worker metrics when the worker is no longer attached after rebalancing occurs. From what I can tell this metric GC code has been in Concourse since v4. 

New theory, maybe something is incorrectly updating when the worker was last seen? I looked through the metrics and found this one: https://github.com/concourse/concourse/blob/49b742951de35f289a5d752f6aa17a6d211f0a6a/atc/metric/metrics.go#L420-L431

In your metrics, do you see `gc-container-collector-dropped` being emitted from web nodes the worker is not attached to? Do you see this metric emitted at all?

**Comment by pn-santos** [comment]
> From what I can tell this metric GC code has been in Concourse since v4.

We only started using metrics with v6.x (I think) but we've had this issue since we started using/tracking metrics.

> In your metrics, do you see `gc-container-collector-dropped` being emitted from web nodes the worker is not attached to?

Humm I assume you mean if I enable the `--emit-to-logs`? Because that particular metric doesn't seem to be emitted by prometheus metrics emitter. It's not mapped in
https://github.com/concourse/concourse/blob/49b742951de35f289a5d752f6aa17a6d211f0a6a/atc/metric/emitter/prometheus.go#L564

which at the end says:
https://github.com/concourse/concourse/blob/49b742951de35f289a5d752f6aa17a6d211f0a6a/atc/metric/emitter/prometheus.go#L645-L646

So from my limited understanding that particular metric isn't published with prometheus (I confirm that no such metric is listed by web instances on our setup).

**Comment by taylorsilva** [comment]
Good catch! I'm not super familiar with the metrics package either. Even though this metric is not emitted it could be updated the `lastSeen` value since that is updated at the top of the function with the giant `switch`:
https://github.com/concourse/concourse/blob/0874fc5969abc560af526b4c812147871654e154/atc/metric/emitter/prometheus.go#L564-L569

Need to test this theory, I'll see if I have time this week.

**Comment by robwhitby** [comment]
Any progress on this? We're seeing it too with `concourse_workers_tasks` and `concourse_workers_containers`. We'd like to be able to rely on these metrics to make scaling decisions.

**Comment by taylorsilva** [comment]
We haven't figured out a reproducible case and looking at the code hasn't yielded any results either. We don't know what the root cause is yet. Current theory is that some metric keeps updating which results in `updateLastSeen()` being called, but even that feels really sketchy.

**Comment by databus23** [comment]
Looking at the metrics code I have a theory why the garbage collection inside the prometheus emitter for worker metrics is not working as expected:

As @taylorsilva mentioned it seems some emitted metric keeps updating the `workerLastSeen` timestamp for workers that are no longer connected to the atc.

From looking at `metrics.go` there are only 6 metrics that carry the `worker` attribute that is used to [trigger an update](https://github.com/concourse/concourse/blob/b852b9dd4003f6bdd2451ec134f9982d8e4b6f12/atc/metric/emitter/prometheus.go#L1304) on the `workerLastSeen` state:

 * `worker containers` - emitted by `workerserver` (register, heartbeat)
 * `worker tasks` - emitted by `workerserver` (register, heartbeat)
 * `worker volumes` - emitted by `workerserver` (register, heartbeat)
 * `worker unknown volumes` - emitted by `volumeserver`
 * `worker unknown containers` - emitted by `containerserver`
 * ~`GC container collector job dropped`~ not used anymore
 
As for as I understand the metrics emitted by the `workerserver` are ok because they are emitted in the `register` and `heartbeat` handler of the worker.
I think the problem are the other two metrics with are emitted by the `containerserver` and `volumeserver` handlers. Both handlers seem to be called via [`tsa.WorkerStatus`](https://github.com/concourse/concourse/blob/8c8eb0565101b287107fe9d36ef970505ad166b7/tsa/tsacmd/remote_cli.go#L268) which seems to be picking a [random ATC](https://github.com/concourse/concourse/blob/8c8eb0565101b287107fe9d36ef970505ad166b7/tsa/tsacmd/command.go#L101). Looks to me like this report request end up on a random atc causing the stale worker metrics.

**Comment by databus23** [comment]
Ok I debugged this some more and I think I finally understand what is happening:

The `worker unknown ...` events emitted by `volumeserver` and `containerserver` are submitted by the worker independently of the worker registration. So while the worker is registered to one tsa/atc via one ssh connection, the volume and container reports are hitting random tsa instances (and etc). 

Because the `lastSeenState` is updated by any metric that carries the `worker` attribute the garbage collection is not triggered and metrics only reported via the registration/heartbeating (`worker containers|tasks|volumes` ) persist as stale metrics.

A simple fix can be to only update the `workerLastSeen` state for metrics that come in via the heartbeating.

I will prepare a pull request for that.
